dotCMS/src/main/java/com/dotcms/security/apps/SecretCachedKeyStoreImpl.java

```diff
@@ -0,0 +1,204 @@
+     */
+    @Override
+    public boolean containsKey(final String variableKey) {
+        try {
+            return getKeysFromCache().contains(variableKey);
+        } catch (Exception e) {
+            Logger.debug(SecretCachedKeyStoreImpl.class,e.getMessage());
+            throw new DotRuntimeException(e);
+        }
+    }
+
+    /**
+     * while in cache it must be encrypted!
+     * once returned from this method it should get decrypted
+     * @param variableKey
+     * @return
+     */
+    @Override
+    public Optional<char[]> getValue(final String variableKey) {
+        final char[] fromCache = getFromCache(variableKey,
+                () -> secretsKeyStore.getValue(variableKey));
+        return Arrays.equals(fromCache, CACHE_404) ? Optional.empty()
+                : Optional.ofNullable(secretsKeyStore.decrypt(fromCache));
+    }
+
+    /**
+     * Lists all keys.
+     * Keys will be loaded from cache if available otherwise they will be loaded from disk
+     * @return
+     */
+    @Override
+    public Set<String> listKeys() {
+        return Sneaky.sneak(this::getKeysFromCache);
+    }
+
+    @Override
+    public boolean deleteAll() throws Exception {
+        secretsKeyStore.destroy();
+        flushCache();
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param variableKey
+     * @param variableValue
+     * @return
+     */
+    @Override
+    public boolean saveValue(final String variableKey, final char[] variableValue) {
+        try {
+            putInCache(variableKey, secretsKeyStore.saveValue(variableKey, variableValue));
+        } catch (DotCacheException e) {
+            Logger.warn(SecretCachedKeyStoreImpl.class, "Unable to save secret in cache " ,e);
+            throw new DotRuntimeException(e.getMessage(), e);
+        }
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param secretKey
+     */
+    @Override
+    public void deleteValue(final String secretKey) {
+        secretsKeyStore.deleteValue(secretKey);
+        try{
+            flushCache(secretKey);
+        } catch (DotCacheException e) {
+            Logger.warn(SecretCachedKeyStoreImpl.class, "Error flushing cache while removing value from secrets store ", e);
+            throw new DotRuntimeException(e);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void backupAndRemoveKeyStore() throws IOException {
+        secretsKeyStore.backupAndRemoveKeyStore();
+    }
+
+    /**
+     * Short hand accessor to get values from cache impl
+     * @param key
+     * @param defaultValue
+     * @return
+     */
+    private char[] getFromCache(final String key, final Supplier<char[]> defaultValue) {
+        return cache.getFromCache(key, defaultValue);
+    }
+
+    /**
+     * Short hand accessor to put values in cache impl
+     * @param key
+     * @param val
+     * @throws DotCacheException
+     */
+    private void putInCache(final String key, final char[] val)
```

Comment:
Codacy found an issue: [Consider using varargs for methods or constructors which take an array the last parameter.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6286472)